### ISSUE TITLE
Input audio and midi at the same time

### ISSUE THREAD

**Issue by jonjamcam** [elieserdejesus/JamTaba#102]
Some plugins/instruments need you to input audio and midi at the same time (vocoders, loopers). Currently there's no way to do that in Jamtaba.

Reaper solves this issue by flexible routing capabilities. So you create a new track in Reaper and send it to where you need.

I was thinking about this lately and there's another big problem. What if you have two or three keyboards and need them all to connect to KONTAKT for example?

Let's try to think in simple terms. You have already a rack to choose plugins with four racks. I use at most two and I need two extra inputs. My question is could this racks be used as extra MIDI inputs when not used by plugins?

Midi has this structure separated in channels, so mixing midi connections does not mean mixing midi information if they are on separate channels. Audio is a different matter. you always need a mixer.

So in short.......same configuration as now, but the first two racks can be used as midi inputs or vstplugins, but not both. what do you think?

**Comment by elieserdejesus** [comment]
@jonjamcam , at moment we have *no input*, *audio*,  and *midi* options in track input panel. I was thinking in add a new option *audio + MIDI*. Seens to me this solve the vocoder issue. What you think about?

But you mentioned the **extra MIDI inputs** too. but seens to me this is a sepparated issue?

**Comment by jonjamcam** [comment]
> add a new option audio + MIDI.

Yeah, it's good enough.

 (Now that I think about it what I wrote previously (https://github.com/elieserdejesus/JamTaba/issues/102#issue-115690497) was too complicated. It's good to think in simpler terms)

**Comment by elieserdejesus** [comment]
@jonjamcam , I realized this feature is complicated from the UI perspective. The problem is how users will select audio and MIDI input? 

I think in some options and at moment the best solution I can imagine is allow audio and midi everytime. So users always will see a midi input selection and an audio input selection. User can select "no input" to both audio and midi, can select valid audio input or valid midi input, or both. 

I think this idea work, but I don't it's good from an end user perspective. The UI will be more complex, and just to support a very exceptional case: AUDIO + MIDI in vocoders, etc. 

I think we need think more about this feature until we have a good solution.

Please tell me if you have another ideas.

**Comment by jonjamcam** [comment]
> the UI will be more complex, and just to support a very exceptional case: AUDIO + MIDI in vocoders, etc.
> Please tell me if you have another ideas.

Sure. No problem. For the moment I have this one idea:

 Use the subchannel as the additional input (midi or audio) adding a button or a checkbox, for example.

So to implement it let's say you select a **vocoder** in Channel 1:

1.- If you select **Audio** as the input, then you add a subchanel, you check the "route subchannel to main button" and it lets you select **midi** ot **no input** only.

2.- If on the other hand you select **Midi** sa the main input, you add a subchannel, check the "button" and it lets you select **Audio** or **no input**.

This feature applies only to standalone because D.A.W. in general have lots of routing options.

So in short, the "button" will route the "other" type of input (midi or audio) not to the master as it is now, but to the main channel.

Do you think this is a better implementation @elieserdejesus?

**Comment by elieserdejesus** [comment]
@jonjamcam , I think this is a good idea. 

Some other things/details to think about before start coding:

1 - use a button in the local track to activate this "hybrid input mode"? Where?  Or a menu entry in the "gear" button, like entries to add and remove channels.

2 - how will be the visual feedback when the 2 subchannels are acting as 1 channel using 2 input types?

**Comment by jonjamcam** [comment]
> use a button in the local track to activate this "hybrid input mode"? Where? Or a menu entry in the "gear" button, like entries to add and remove channels.

1.- In the spirit of trying to use what's already there it occured to me that we could use a "tools button" type for this function:

![image](https://cloud.githubusercontent.com/assets/15310433/22351975/02eeb1b0-e3f9-11e6-89bd-a472a9c71673.png)

2.- I like your idea of using the gear button too. It's clear to have a separate entry that says "add slave subchannel" or something along these lines.

**Comment by elieserdejesus** [comment]
@jonjamcam , I like your "check box" idea in Midi tools dialog. It's very clear. We can add a "gear" menu entry too, available only when user created the second subchannel.

I think is better (from user perspective) allow **midi routing** only in second subchannel. The steps to use a vocoder, for example, will be:
1 - insert the vocoder plugin in the first subchannel
2 - select audio input in first channel
3 - create the second subchannel
4 - select MIDI input
5 - open MIDI tools dialog and check the **route to main** checkbox.

In my mind make more sense always create the audio subchannel first, and route the midi track. I can't remember plugins where make more sense create the midi track first and route the audio track . Can you imagine/remember some plugns where is better create midi subchannel first?

The arrow idea for visual feedback is perfect. I'm just a bit worried with track layout, if we put a icon on top the vertical align will change, but we can think in something, it's not a complex issue.

**Comment by jonjamcam** [comment]
> I can't remember plugins where make more sense create the midi track first and route the audio track . Can you imagine/remember some plugns where is better create midi subchannel first?

I agree. In my experience where I need to have both input types, normally I route the audio first and then the midi as secondary input for control purposes (vocoder, audio looper).

**Comment by jonjamcam** [comment]
So if this feature will be used for midi only, how do we handle the vst racks and the solo button? Should they be disabled in this particular case?

 Theres a possibility that you insert a vst plugin that sends audio instead of midi, so in this case the feature may not work as expected.

**Comment by elieserdejesus** [comment]
@jonjamcam , I created a PR to this feature: https://github.com/elieserdejesus/JamTaba/pull/682

> So if this feature will be used for midi only, how do we handle the vst racks, the solo button, gain buttons and the stereo invert button of the second subchannel? Should they be disabled in this particular case?

I think it's a good idea disable these buttons, will be clear for users.

> Theres a possibility that you insert a vst plugin that sends audio instead of midi, so in this case the feature may not work as expected.

At moment is **midi routing only**.

**Comment by jonjamcam** [comment]
> At moment is midi routing only.

Yes that's perfect. I don't think audio routing is necessary. My question is what would happen if you insert a vst plugin that sends audio instead of midi, as kontak for example. Do you think @elieserdejesus this could mess the routing?

**Comment by elieserdejesus** [comment]
> ... as kontak for example. Do you think @elieserdejesus this could mess the routing?

I created a task in the PR to handle this situation, but I'm not shure what is the best option.  Let's consider an example:

1 - Select audio input in first subchannel
2 - Create e 2nd subchannel and select midi input
3 - Activate the midi routing in MIDI tools dialog (working as expected).
4 - Now, **add Kontakt in 2nd subchannel**. What is the expected midi input behavior now? The midi input will be routed to first subchannel ONLY? Kontakt will see the midi messages? 

Is possible send the midi messages to both routed channel and kontakt, but **I think it's more clear send only to the routed subchannel**. In my mind when we check the "MIDI routing" option we are thinking in **change/redirect the MIDI route** and not in "add a second MIDI route". 

What you think @jonjamcam , what is more clear for you?

**Comment by jonjamcam** [comment]
> In my mind when we check the "MIDI routing" option we are thinking in change/redirect the MIDI route and not in "add a second MIDI route".

Yes. I agree.

**Comment by jonjamcam** [comment]
> Let's consider an example:
>
>1 - Select audio input in first subchannel
2 - Create e 2nd subchannel and select midi input
3 - Activate the midi routing in MIDI tools dialog (working as expected).
4 - Now, add Kontakt in 2nd subchannel.

Same example, don't you think @elieserdejesus that the vst racks should be disabled in the first place? I don't see a point in having vst plugins for this function. Same with fader and pan.

**Comment by elieserdejesus** [comment]
@jonjamcam , are you suggesting disable plugins in the second subchannel, right? If yes, I agree is better disable all audio related controls (pan, fader, etc).

**Comment by jonjamcam** [comment]
> is better disable all audio related controls (pan, fader, etc).

Yes.

**Comment by elieserdejesus** [comment]
This issue was implemented in PR #682.  Thanks for your help and ideas @jonjamcam !

**Comment by jonjamcam** [comment]
Thanks to you @elieserdejesus for considering this feature.
